lib/using_restricted_array.rb

```diff
@@ -6,58 +6,119 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def length(array)
-  raise NotImplementedError
+  index = 0
+  while array[index] != nil do
+    index += 1
+  end
+  return index
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def print_array(array)
-  raise NotImplementedError
+  index = 0
+  while array[index] != nil do
+    puts array[index]
+    index += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  raise NotImplementedError if length == 0
+  i = 0
+  while i < length do
+    if array[i] == value_to_find
+      return true
+    end
+    i += 1
+  end
+  return false
 end
 
+
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  raise NotImplementedError if length == 0
+  max_length = 0
```

Comment:
This assumes that the values are all positive.  I suggest this change

```suggestion
  max_length = array[0]
```